primer-miso/src/Primer/Miso/Util.hs

```diff
@@ -250,3 +286,89 @@ nodeSelectionType =
           _ -> False
+
+-- TODO keep an eye out for whether these ever return `Nothing` in practice
+-- the current hypothesis is that, because we already typecheck on initialisation and on performing each action,
+-- the app only ever deals with typechecked expressions
+-- and that it is therefore possible that we could refactor the library to assert this in the Haskell types
+-- note that we try to always work with typechecked expressions in the frontend,
+-- as being able to assume that metadata is always available makes the core application code simpler
+assumeDefHasTypeCheckInfo :: ASTDef -> Maybe ASTDefT
+assumeDefHasTypeCheckInfo def = do
+  expr <- sequenceOf (_exprMeta % _type) (astDefExpr def) >>= sequenceOf (_exprTypeMeta % _type)
+  sig <- sequenceOf (_typeMeta % _type) (astDefType def)
+  pure ASTDefT{expr, sig}
+assumeDefSelectionHasTypeCheckInfo :: DefSelection (Either ExprMeta (Either TypeMeta KindMeta)) -> Maybe DefSelectionT
+assumeDefSelectionHasTypeCheckInfo =
+  traverseOf #node $ traverse $ traverseOf #meta $ bitraverse (sequenceOf _type) $ bitraverse (sequenceOf _type) pure
+
+-- see `assumeDefHasTypeCheckInfo`
+-- sometimes we need to discard that extra type-level information,
+-- in order to get an input for various Primer library functions
+astDefTtoAstDef :: ASTDefT -> ASTDef
+astDefTtoAstDef def = ASTDef{astDefExpr = exprTtoExpr def.expr, astDefType = typeTtoType def.sig}
+
+-- this is potentially a better API then the one which `Primer.Action.Available` currently exports
+availableForSelection ::
+  HasID a =>
+  TypeDefMap ->
+  DefMap ->
+  Level ->
+  Editable ->
+  ASTDef ->
+  DefSelection a ->
+  [Action]
+availableForSelection tydefs defs level editable def defSel = case defSel.node of
+  Nothing -> Available.forDef defs level editable defSel.def
+  Just nodeSel -> case nodeSel.nodeType of
+    BodyNode -> Available.forBody tydefs level editable (astDefExpr def) (getID nodeSel)
+    SigNode -> Available.forSig level editable (astDefType def) (getID nodeSel)
+
+-- this part of the actions API needs a re-think
+-- (it was perhaps too motivated by what was convenient for our old TypeScript frontend):
+-- `context` is a bit weird, and basically just means module (lack of context can mean local name or primitive)
+-- `optToName` is used for presented choices - we know these are variables, not chars or ints
+-- use of `unsafeMk` functions is necessary even though we need never convert from names to text in the first place
+-- `matchesType` is meaningless in `stringToOpt` - we should use different types for presented options and submissions
+optToName :: Available.Option -> Either (LocalName l) (GlobalName g)
+optToName opt =
+  maybe
+    (Left . unsafeMkLocalName)
+    (curry $ Right . unsafeMkGlobalName)
+    opt.context
+    opt.option
+stringToOpt :: Text -> Available.Option
+stringToOpt t = Available.Option t Nothing True
+
+-- should DRY with function of some name in `primer-api`
```

Comment:
I think you mean "same name" here. Anyway, probably worth adding to the TODO list.